### PR TITLE
reset password notification

### PR DESCRIPTION
**Pull request by kostyrin**

### CODE REVIEW

File: VirtoCommerce.Platform.Core/Notifications/PlatformEmailSender.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace VirtoCommerce.Platform.Core.Notifications
+{
+    public class PlatformEmailSender : IEmailSender
+    {
+        public Task SendEmailAsync(string to, string type, string languageCode, Dictionary<string, object> parameters)
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Core/Notifications/PlatformEmailSender.cs]: Use this signature  `public Task SendEmailAsync(string email, string subject, string message)`

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Model/NotificationParameterAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace VirtoCommerce.NotificationsModule.Core.Model
+{
+    [AttributeUsage(AttributeTargets.Property)]
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Model/NotificationParameterAttribute.cs]: This attribute redundant, we should avoid of using any descriptions at the code level.
For  an available properties  legend  we use JSON resources in UI

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Types/ResetPasswordEmailNotification.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+using VirtoCommerce.NotificationsModule.Core.Model;
+
+namespace VirtoCommerce.NotificationsModule.Core.Types
+{
+    public class ResetPasswordEmailNotification : EmailNotification
+    {
+        [NotificationParameter("Reset password URL")]
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Types/ResetPasswordEmailNotification.cs]: This attribute redundant, we should avoid of using any descriptions at the code level.
For an available properties legend we use JSON resources in UI

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Senders/NotificationEmailSender.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            {
+                emailNotification.To = to;    
+            }
+
+            ParametersResolver(notification, parameters);
+
+            await _notificationSender.SendNotificationAsync(notification, languageCode);
+        }
+
+        private void ParametersResolver(Notification notification, Dictionary<string, object> parameters)
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Data/Senders/NotificationEmailSender.cs]: This is also redundant

File: VirtoCommerce.Platform.Core/Notifications/IEmailSender.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace VirtoCommerce.Platform.Core.Notifications
+{
+    public interface IEmailSender
+    {
+        Task SendEmailAsync(string to, string type, string languageCode, Dictionary<string, object> parameters);
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Core/Notifications/IEmailSender.cs]: Use this signature  `public Task SendEmailAsync(string email, string subject, string message)`

File: VirtoCommerce.Platform.Core/Extensions/DIExtension.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+using System.Linq;
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Core/Extensions/DIExtension.cs]: Why do you need this extension?  Could you explain?

Comment (kostyrin) [reply VirtoCommerce.Platform.Core/Extensions/DIExtension.cs]: There are two implementation: PlatformEmailSender(with NotImplementException) and NotificationEmailSender(with true implementation) in the DI container. If we install the NotificationsModule then we use  NotificationEmailSender instead of PlatformEmailSender. So, we replace the implementation. This one of kind replacing of an implementation.

File: VirtoCommerce.Platform.Web/appsettings.json [hunk 7]
```diff
@@ -18,6 +18,7 @@
     },
 
     "EmailSendingOptions": {
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Web/appsettings.json]: We should define module specific settings within individual configuration section for each module. 
"Notification" for  Notification module

File: VirtoCommerce.Platform.Web/Startup.cs [hunk 8]
```diff
@@ -65,9 +61,12 @@ public void ConfigureServices(IServiceCollection services)
 
             services.Configure<DemoOptions>(Configuration.GetSection("VirtoCommerce"));
             services.Configure<HangfireOptions>(Configuration.GetSection("VirtoCommerce:Jobs"));
+            services.Configure<EmailSendingOptions>(Configuration.GetSection("EmailSendingOptions"));
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Web/Startup.cs]: This line shouldn't be here!!!

File: VirtoCommerce.Platform.Web/Controllers/Api/SecurityController.cs [hunk 9]
```diff
@@ -409,11 +413,12 @@ public ActionResult GetAllRegisteredPermissions()
...
-            if (user != null && user.Email != null && IsUserEditable(user.UserName) && !(await _userManager.IsInRoleAsync(user, SecurityConstants.Roles.Customer)))
+            if (user != null && user.Email != null /*&& IsUserEditable(user.UserName) && !(await _userManager.IsInRoleAsync(user, SecurityConstants.Roles.Customer))*/)
             {
                 var token = await _userManager.GeneratePasswordResetTokenAsync(user);
                 var callbackUrl = $"{Request.Scheme}{Request.Host}/api/platform/security/#/resetpassword/{user.Id}/{token}";
 
+                await _emailSender.SendEmailAsync(user.Email, "Reset password",  callbackUrl );
                 //TODO: Generate Domain Event and implement the sending password reset email to user in event handler
 
                 //var notification = _notificationManager.GetNewNotification<ResetPasswordEmailNotification>("Platform", typeof(ResetPasswordEmailNotification).Name, "en");
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Web/Controllers/Api/SecurityController.cs]: Remove this commented line

File: Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Types/ResetPasswordEmailNotification.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+using VirtoCommerce.NotificationsModule.Core.Model;
+
+namespace VirtoCommerce.NotificationsModule.Core.Types
+{
+    public class ResetPasswordEmailNotification : EmailNotification
```
Comment (tatarincev) [inline Modules/vc-module-notifications/VirtoCommerce.NotificationsModule.Core/Types/ResetPasswordEmailNotification.cs]: Missed comments for this  notification type

File: VirtoCommerce.Platform.Core/Notifications/EmailSendingOptions.cs [hunk 11]
```diff
@@ -1,10 +1,15 @@
-namespace VirtoCommerce.NotificationsModule.Core.Model
+namespace VirtoCommerce.Platform.Core.Notifications
 {
     /// <summary>
     /// Settings of Email Protocol
     /// </summary>
     public class EmailSendingOptions
     {
+        /// <summary>
+        /// Email address of The Sender
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Core/Notifications/EmailSendingOptions.cs]: This property should be called 'DefaultSender'

File: VirtoCommerce.Platform.Core/Notifications/EmailSendingOptions.cs [hunk 12]
```diff
@@ -1,10 +1,15 @@
-namespace VirtoCommerce.NotificationsModule.Core.Model
+namespace VirtoCommerce.Platform.Core.Notifications
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Core/Notifications/EmailSendingOptions.cs]: EmailSendingOptions should be defined in Notifications module, why did you move it to the platform?

File: VirtoCommerce.Platform.Core/Extensions/DIExtension.cs (hunk 6 above)
Comment (kostyrin) [reply VirtoCommerce.Platform.Core/Extensions/DIExtension.cs]: It is working, I cheked

File: VirtoCommerce.Platform.Core/Notifications/EmailSendingOptions.cs (hunk 12 above)
Comment (kostyrin) [reply VirtoCommerce.Platform.Core/Notifications/EmailSendingOptions.cs]: Agree, there was the options in NotificationsModule in the first version. That's mistake did after talking: Андрей в strartup and etc.

File: VirtoCommerce.Platform.Web/appsettings.json [hunk 13]
```diff
@@ -19,16 +19,23 @@
...
 
-    "EmailSendingOptions": {
-        "SmtpOptions": {
-            "SmtpServer": "",
-            "Port": 587,
-            "Login": "",
-            "Password": ""
+    "Notifications": {
+        "Gateway": "Default",
+        "Default": {
```
Comment (tatarincev) [inline VirtoCommerce.Platform.Web/appsettings.json]: Better will define two different Options for each sender i.e `SmtpSenderOptions` and `SendGridSenderOptions`  insead of generic `EmailSendingOptions`
and use follow syntax in appsettings.json
```
Notifications : {
Gateway : "Smtp|SendGrid"
.....
}
"SendGrid" : {  ... },
"Smtp" : { ... }
```
